Added capability to configure TestIds to use ranges for Softbins

priyavadan opened the pull request:
TestIds can now be configured to accept ranges for softbins

example interface setup for TestIds

~~~ruby
        TestIds.configure :wafer_test do |config|
          config.softbins.size = 5
          config.softbins do |bin, options|
           if !options[:softbin].nil? && options[:softbin].is_a?(Range) 
            TestIds.next_in_range(bin, options)
           end
          end
          config.numbers = :sssssxx
          config.numbers.size = 2
        end
~~~

From the flow or interface, the soft bins can be supplied as ranges.

Example:

~~~ruby
       case options[:bin]
          when 100
           options[:softbin] = (1000..2000)
          when 101
           options[:softbin] = (2001..3000)
          when 102
           options[:softbin] = (4000..5000)
          when 103
           options[:softbin] = (8000..10000)
...

or 

      options[:softbin] = (5000..6000)
~~~

P.S: also lint fixes

Commit:
Added Soft Bin Range capability, Initial check-in

Commit:
Pointer for sofbin ranges now increments based on softbin size

Commit:
Changed given to assigned inside the next in range method

Commit:
Added check for assigned softbin to be within range, else fail with an error

Commit:
Few more changes

Commit:
Added comments

Commit:
removed redundant option

Commit:
reordered test_ids

Commit:
corrected typo (global search had replaced)

info-rchitect commented:
@priyavadan very cool!  any possibility to extend this to bin as well?  I was thinking similar to [live updating parameters](http://origen-sdk.org/origen/guides/models/parameters/#Live_Updating_Parameters) in concept.

Commit:
Fixed re-ordering, method needs to be publicc

priyavadan commented:
@info-rchitect yes, I think it can be extended. 

Thanks for the link, I'll take a look at the live updating parameters concept later today before commenting further.

info-rchitect commented:
@priyavadan for us, the need is to assign hardbins per test type for some of our test modules.  Many have one bin for all tests in the test module but there are a few exceptions.  For example, separate hardbins for opens vs shorts at FT or separate hardbins for scan chain, logic, and delay at WS.

priyavadan commented:
@info-rchitect ok, I can help guide you in the right direction for get the plugin updated. Since you understand the requirements it would be better for you to take a stab at it.

My need for this PR to be approved is urgent as it is/will gate development on a new product.

info-rchitect commented:
@priyavadan works for me, thx for the help!

Commit:
added myself to dev mail list

ginty commented on lib/test_ids.rb:
```diff
@@ -15,156 +15,160 @@ module TestIds
-      end
-      @publish = val ? :save : :dont_save
-    end
+   # Allocates a number to the given test and returns a new hash containing
```
What's caused this diff, editing in Windows or something? Can you get rid of it?

ginty commented on lib/test_ids/allocator.rb:
```diff
@@ -14,6 +14,72 @@ def initialize(configuration)
+        @pointer = 0
+        rangehash.merge!("#{range}": "#{range.to_a[@pointer]}")
+        assigned_softbin = range.to_a[@pointer]
+        store['pointers']['softbin'] = assigned_softbin
```
Should not be updating this pointer, this is the pointer that is used by TestIDs whenever softbin is configured like:

~~~ruby
config.softbins.include << (1000..2000)
~~~

You should be creating a separate pointer here that lives under `store['pointers']['ranges'][<range_identifier>]`, possibly also with a key if you want to support the same range being used independently for two different purposes (e.g. bin and softbin) - `store['pointers']['ranges'][<key>][<range_identifier>]`

ginty commented on lib/test_ids/allocator.rb:
```diff
@@ -14,6 +14,72 @@ def initialize(configuration)
+    # The configuration for the TestId plugin needs to pass in the bin number and the options from the test flow
+    # For this method to work as intended.
+    def next_in_range(softbin, options)
+      unless options.nil?
```
I would expect this method to be dealing with returning either a next in line number, or else calling a reclaim method

ginty commented on lib/test_ids/configuration.rb:
```diff
@@ -92,6 +92,10 @@ def bins
       @bins ||= Item.new
```
The stuff in here doesn't seem to actually be used

ginty replied:
Still doesn't seem to be used

ginty commented:
Hi @priyavadan, thanks for the PR and which is definitely adding welcome functionality.

Some work is required to make the implementation more generic, the `next_in_range` feature should not be coupled to the softbin allocation at all, it should be a separate and generic feature thereby allowing it to be used in a callback function to assign any number type.

In your example:

~~~ruby
TestIds.configure :wafer_test do |config|
  config.softbins.size = 5
  config.softbins do |bin, options|
    if !options[:softbin].nil? && options[:softbin].is_a?(Range) 
      TestIds.next_in_range(bin, options)
    end
  end
  config.numbers = :sssssxx
  config.numbers.size = 2
end
~~~

it is already going down a wrong path by passing the softbin as an option into TestIds (also I don't understand why the (hard) bin is passed in?), that should be extracted at the app level so that only a generic range is passed in:

~~~ruby
TestIds.configure :wafer_test do |config|
  config.softbins.size = 5
  config.softbins do |bin, options|
    if options[:softbin].is_a?(Range) 
      TestIds.next_in_range(options[:softbin])
    end
  end
  config.numbers = :sssssxx
  config.numbers.size = 2
end
~~~

Then you naturally have something that can be used in any callback:

~~~ruby
config.bins do |options|
  if options[:scan_test]
    TestIds.next_in_range(0..9999)
  else
    TestIds.next_in_range(10000..20000)
  end
end
~~~

Commit:
fix 1

Commit:
some updates as per feedback from @ginty

priyavadan commented:
Thanks @ginty !! Appreciate the inputs.

I have made some changes and updates that I think achieve the vision you have in mind. Please let me know if I did not understand it correctly!

Commit:
fixed typo in comments

ginty commented on lib/test_ids/allocator.rb:
```diff
@@ -411,6 +475,7 @@ def reclaim_bin(options)
     def allocate_softbin(options)
       bin = options[:bin]
       num = options[:number]
+      softbin = options[:softbin] if options[:softbin]
```
Should not be required

priyavadan replied:
ok, removed.

Commit:
updated as per feedback

priyavadan commented:
Sorry, if I was unclear before. 

I am proposing that these changes be merged for now if the updates to this PR are acceptable. 

I will work on adding the reclaim for ranges in a separate PR.

ginty commented on lib/test_ids/allocator.rb:
```diff
@@ -14,6 +14,63 @@ def initialize(configuration)
+    # The configuration for the TestId plugin needs to pass in the bin number and the options from the test flow
+    # For this method to work as intended.
+    def next_in_range(range, options)
+      unless options.nil?
```
Why is this gate here?

priyavadan replied:
left over from previous implementation, will remove

ginty commented on lib/test_ids/allocator.rb:
```diff
@@ -14,6 +14,63 @@ def initialize(configuration)
+      # Create an alias for the databse that stores the pointers per range
+      rangehash = store['pointers']['ranges'] ||= {}
+      # Setup an identifier to store in the database, dont really care if it is a bin, softbin or number at this point.
+      identifier = "last_assigned_value_from_given_range"
```
Why is there a hardcoded identifier? What purpose is it serving if the user cannot provide an option to override it?

ginty commented on lib/test_ids/allocator.rb:
```diff
@@ -14,6 +14,63 @@ def initialize(configuration)
+      store['pointers'][identifier] ||= [] 
+      # Not really using the Item.new currently, but instantiating the BinArray regardless
+      # for future enhancement/use.
+      r ||= TestIds::Configuration::Item.new
```
Would be better not to do it then

priyavadan replied:
Will remove.

ginty commented on lib/test_ids/allocator.rb:
```diff
@@ -14,6 +14,63 @@ def initialize(configuration)
+        # Now update the database pointers to point to the lastest assigned softbin for a given range.
+        rangehash.merge!("#{range}": "#{range.to_a[@pointer]}")
+        # Also update the database so that the latest given softbin across all ranges is the latest.
+         store['pointers'][identifier] = assigned_value
```
Why, what is this used for?

priyavadan replied:
I am currently not using it and will remove.

ginty commented on lib/test_ids/allocator.rb:
```diff
@@ -109,6 +166,11 @@ def allocate(instance, options)
         softbin['number'] ||= allocate_softbin(bin: bin['number'], number: number['number'], size: softbin_size)
         softbin['size'] ||= softbin_size
+      elsif options[:softbin].is_a?(Range)
+        softbin['number'] ||= allocate_softbin(bin: bin['number'], softbin: options[:softbin], size: softbin_size)
```
What is this code for?

priyavadan replied:
I'll attempt to explain.

in this code, the else condition is the one that gets executed because
~~~
config.softbins.algorithm && config.softbins.algorithm.to_s =~ /n/ = nil 
(config.softbins.callback && !config.numbers.function?) = false 
~~~

With the code statement below

~~~
softbin['number'] ||= allocate_softbin(bin: bin['number'], size: softbin_size)
~~~

We are just passing the bin and softbin size as options to the allocate_softbin method, which in turn calls the callback.

~~~
callback.call(bin, options)
~~~

The options at this time contain only the bin and size that we passed in before.

Passing just the options to allocate_softbin will not work either because size is used in the range_item method to increment the pointer by the size configured for either bin/softbin/number.

~~~
From range_item method:

       if previous_assigned_value == range.to_a[@pointer]
          @pointer += options[:size]
          assigned_value = range.to_a[@pointer]
        else
          # Because of the pointer calculations above, I don't think it will ever reach here, has not in my test cases so far!
          assigned_value = range.to_a[@pointer]
        end
~~~

ginty replied:
This should not be required, but I guess it is working around the fact that the application is passing in something like options[:softbin] == (1000..2999) and this code is trying to work around the existing convention that any softbin option passed in here will take precedence and it will not fall through the to callback.

This should be the application's responsibility to resolve, so actually your example code should really be:

~~~ruby
  config.softbins do |bin, options|
    if options[:softbin_range] 
      TestIds.next_in_range(options[:softbin_range])
    end
  end
~~~

i.e. it should not used the reserved option :softbin to pass the range about.

ginty reviewed (changes requested):
Sorry, but there seems to be a lot of cruft in here which really makes me doubt this solution overall.
Please also add test cases to show that it works and provide test coverage.
A doc update is required too.
Thanks.

Commit:
more updates

Commit:
Further updates, added spec case

Commit:
Removed ranges

Commit:
Fixed typo

Commit:
allocate(bin/softbin/number) methods should now have access to all flow options

Commit:
Removed unwanted code

Commit:
more cleanup

Commit:
Lint fix

Commit:
Added Documentation

ginty commented on lib/test_ids/allocator.rb:
```diff
@@ -460,8 +508,15 @@ def allocate_softbin(options)
+      #
+      # elsIf softbins is a callback and number does reference softbin
+      #   callback.call(bin, options)
+      #
```
Why is this comment here if not implemented?

priyavadan replied:
Will remove, it(the logic explained in the comments) is not required anymore.

ginty commented on lib/test_ids/allocator.rb:
```diff
@@ -98,21 +146,21 @@ def allocate(instance, options)
+        softbin['number'] ||= allocate_softbin(options.merge(bin: bin['number'], size: softbin_size))
         softbin['size'] ||= softbin_size
-        number['number'] ||= allocate_number(bin: bin['number'], softbin: softbin['number'], size: number_size)
+        number['number'] ||= allocate_number(options.merge(bin: bin['number'], softbin: softbin['number'], size: number_size))
```
Thanks for adding the options to all of these method calls, but the feature is only half-implemented if the corresponding calls to the callbacks don't pass the options - i.e. you have done it for the softbin callback you care about, why not the rest?

priyavadan replied:
Will add and enable passing in specific ranges for bins as well with a callback.

Separate set of spec tests have been added to check for different configuration possibilities.

Commit:
- Added separate spec tests for different configurations possible when
using specific ranges
- bins, numbers and softbins now can be generated by specifying ranges for
each test inside the configuration.
- For examples look at the spec/specific_ranges_spec.rb file.
- The constraints at http://origen-sdk.org/test_ids/#Configuration need
  to be remembered when specifying ranges.
- Documentation was updated, added some examples.

Commit:
Added another line in documentation

priyavadan commented on lib/test_ids/allocator.rb:
```diff
@@ -427,9 +427,14 @@ def add_missing_references(config_item, references)
+    # Not sure if this is the right way. IMO the following are true: 
+    # 1. config.bins will have a callback only when ranges are specified.
+    # 2. If config.bins is empty but config.bins is not a callback, return nil to maintain functionality as before.
+      return nil if config.bins.empty? && !config.bins.callback  
```
@ginty please comment if this logic change is acceptable or not. Bins will be generated via callback if a user specifies specific bin ranges in the configuration.

relevant example in spec/specific_ranges_spec.rb

ginty replied:
Looks good to me

Commit:
corrected new file structure description

ginty committed:
Enabled CI

ginty committed:
Lint cleanup

ginty reviewed (approved):
OK, this is a good feature addition, thanks for the updates!

priyavadan commented:
@ginty  Thank you for the guidance. Will make sure future PR's meet the same standards.

info-rchitect commented:
@priyavadan nice work on the PR and good mentoring by @ginty.  Next up, how to set hardbins based on test metadata!
